### PR TITLE
Update core contract

### PR DESCRIPTION
**Pull request by benjaminbollen**
I will continue from here

### CODE REVIEW

File: contracts/core/Core.sol [hunk 1]
```diff
@@ -221,14 +210,14 @@ contract Core is MasterCopyNonUpgradable, ConsensusModule, MosaicVersion, CoreSt
...
-    uint256 public precommitClosureBlockHeight;
+    /** Precommitment closure block number */
+    uint256 public precommitClosureBlockNumber;
 
     /**
      * Block number at which core status is changed to creation.
      * It will be used by validators to create GenesisFile.
      */
-    uint256 public rootOriginObservationBlockHeight;
+    uint256 public rootOriginObservationBlockNumber;
```
Comment (benjaminbollen) [inline contracts/core/Core.sol:220]: `genesisOriginObservationBlockNumber`

File: contracts/core/Core.sol [hunk 2]
```diff
@@ -35,6 +36,13 @@ contract Core is MasterCopyNonUpgradable, ConsensusModule, MosaicVersion, CoreSt
     /** Emitted when new metablock is proposed */
     event MetablockProposed(bytes32 proposal);
 
+    /** Emitted when core status is changed. */
+    event CoreStatusUpdated(CoreStatus status);
+
+    /**  Emitted when core is opened. */
+    event GenesisOriginObservationStored(uint256 originObservationBlockNumber);
```
Comment (benjaminbollen) [inline contracts/core/Core.sol:43]: `event GenesisOriginObservationStored(uint256 genesisOriginObservationBlockNumber);`

File: contracts/validator-set/ValidatorSet.sol [hunk 3]
```diff
@@ -285,4 +286,15 @@ contract ValidatorSet is ForwardValidatorSetAbstract {
         validatorCount[nextHeight] = validatorCount[nextHeight].sub(1);
         fvsCount[activeHeight] = fvsCount[activeHeight].sub(1);
     }
+
+    /**
+     * @notice It returns current active height.
+     *
+     * @returns Value of active height.
+     */
+    function getActiveHeightInternal() internal view returns(uint256) {
```
Comment (benjaminbollen) [inline contracts/validator-set/ValidatorSet.sol:295]: no no no; this doesnt need to be read

File: contracts/core/Core.sol [hunk 4]
```diff
@@ -355,7 +344,7 @@ contract Core is MasterCopyNonUpgradable, ConsensusModule, MosaicVersion, CoreSt
         minimumValidatorCount = _minValidators;
         joinLimit = _joinLimit;
 
-        creationKernelHeight = _height;
+        creationKernelHeight = getActiveHeightInternal();
```
Comment (benjaminbollen) [inline contracts/core/Core.sol:347]: nooo way

File: contracts/core/Core.sol [hunk 5]
```diff
@@ -722,17 +715,18 @@ contract Core is MasterCopyNonUpgradable, ConsensusModule, MosaicVersion, CoreSt
             uint256 minValidatorCount_,
             uint256 beginHeight_
         )
-    {
+    {   // change to blocknumber
```
Comment (benjaminbollen) [inline contracts/core/Core.sol:718]: forgotten comment

File: contracts/core/Core.sol [hunk 6]
```diff
@@ -742,10 +736,14 @@ contract Core is MasterCopyNonUpgradable, ConsensusModule, MosaicVersion, CoreSt
             );
             coreStatus = CoreStatus.opened;
             // Core status would be changed to `opened`.
-            // So `rootOriginObservationBlockHeight` would be set once.
-            rootOriginObservationBlockHeight = block.number;
+            // So `rootOriginObservationBlockNumber` would be set once.
+            rootOriginObservationBlockNumber = block.number;
```
Comment (benjaminbollen) [inline contracts/core/Core.sol:740]: rename

File: contracts/protocore/SelfProtocore.sol [hunk 7]
```diff
@@ -24,6 +24,10 @@ import "../validator-set/ValidatorSet.sol";
  */
 contract SelfProtocore is MasterCopyNonUpgradable, GenesisSelfProtocore, Protocore, ValidatorSet {
 
+    /* Usings */
+    using SafeMath for uint256;
```
Comment (benjaminbollen) [inline contracts/protocore/SelfProtocore.sol:28]: good catch

File: contracts/test/protocore/TestSelfProtocore.sol [hunk 8]
```diff
@@ -122,6 +122,22 @@ contract TestSelfProtocore is SelfProtocore {
         return links[_voteMessageHash].fvsVoteCount[_height];
     }
 
+    function forwardValidatorCount(uint256 _height)
```
Comment (benjaminbollen) [inline contracts/test/protocore/TestSelfProtocore.sol:125]: best removed?

File: test/core/join.js [hunk 9]
```diff
@@ -51,8 +51,8 @@ contract('Core::join', async (accounts) => {
       epochLength: new BN(100),
       minValidatorCount: new BN(5),
       validatorJoinLimit: new BN(20),
-      height: new BN(1),
-      parent: CoreUtils.randomSha3(),
```
Comment (benjaminbollen) [inline test/core/join.js:55]: unsure what s the thinking here

File: test/core/join_before_open.js [hunk 10]
```diff
@@ -240,7 +240,7 @@ contract('Core::joinBeforeOpen', async (accounts) => {
       const validator = accountProvider.get();
       expectedUpdatedValidators.push(validator);
       expectedUpdatedReputations.push(new BN(1));
-      const expectedRootOriginObservationBlock = await web3.eth.getBlockNumber() + 1;
+      const expectedRootOriginObservationBlockNumber = await web3.eth.getBlockNumber() + 1;
```
Comment (benjaminbollen) [inline test/core/join_before_open.js:243]: `expectedGenesisOriginObservationBlockNumber`

File: test/core/join_before_open.js [hunk 11]
```diff
@@ -292,10 +292,10 @@ contract('Core::joinBeforeOpen', async (accounts) => {
         isProposalSetInitialized,
       );
 
-      const actualRootOriginObservationBlock = await config.mockCore
-        .rootOriginObservationBlockHeight.call();
+      const actualRootOriginObservationBlockNumber = await config.mockCore
```
Comment (benjaminbollen) [inline test/core/join_before_open.js:295]: `actualGenesisOriginObservationBlockNumber`

File: test/core/logout.js [hunk 12]
```diff
@@ -51,8 +51,8 @@ contract('Core::logout', async (accounts) => {
       epochLength: new BN(100),
       minValidatorCount: new BN(5),
       validatorJoinLimit: new BN(20),
-      height: new BN(1),
-      parent: CoreUtils.randomSha3(),
```
Comment (benjaminbollen) [inline test/core/logout.js:55]: again, unsure what s the thinking here

File: test/self_protocore/upsert_validator.js [hunk 13]
```diff
@@ -76,7 +76,6 @@ contract('SelfProtocore::upsertValidator', (accounts) => {
     config.openKernelHeight = new BN(20);
     validator.address = accountProvider.get();
     validator.beginHeight = config.openKernelHeight.addn(1);
-    validator.endHeight = new BN(100);
```
Comment (benjaminbollen) [inline test/self_protocore/upsert_validator.js:79]: what are you doing changing these tests when you re touching core.sol?

**Review by benjaminbollen — COMMENTED** [review]
leaving comments, for mine and your reference.

File: contracts/core/Core.sol [hunk 14]
```diff
@@ -722,17 +715,18 @@ contract Core is MasterCopyNonUpgradable, ConsensusModule, MosaicVersion, CoreSt
...
+        insertValidatorInternal(_validator, creationKernelHeight);
 
         Kernel storage creationKernel = kernels[creationKernelHeight];
         countValidators = creationKernel.updatedValidators.push(_validator);
         // TASK: reputation can be uint64, and initial rep set properly
         creationKernel.updatedReputation.push(uint256(1));
         if (countValidators >= minimumValidatorCount) {
             quorum = calculateQuorum(countValidators);
-            precommitClosureBlockHeight = CORE_OPEN_VOTES_WINDOW;
+            precommitClosureBlockNumber = CORE_OPEN_VOTES_WINDOW;
```
Comment (benjaminbollen) [inline contracts/core/Core.sol]: what was this line in the first place????????? 
????
????

Comment (benjaminbollen) [inline contracts/core/Core.sol:729]: this seems wrong; should have been `block.number + CORE_OPEN_VOTES_WINDOW` no?

File: test/self_protocore/upsert_validator.js (hunk 13 above)
Comment (benjaminbollen) [reply test/self_protocore/upsert_validator.js:79]: I deleted this file; just redo it proper

**Comment by pgev** [comment]
Closing the PR as it's continued here: #258

**Pull request closed without merging**